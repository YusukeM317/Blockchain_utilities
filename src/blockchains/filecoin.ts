import {BlockchainHandler, BlockchainHandlerOpts} from "../blockchain-handler";
import {AccountID} from "caip";
import {ConsentMessage, getConsentMessage, LinkProof} from "../utils";
import type { MessageParams } from '@zondax/filecoin-signing-tools'

const namespace = 'fil'

const moduleToImport = process.env.JEST_WORKER_ID ? "@zondax/filecoin-signing-tools/nodejs" : "@zondax/filecoin-signing-tools"

function asTransaction(address: string, message: string): MessageParams {
    const messageHex = Buffer.from(message).toString('hex')
    return {
        from: address,
        to: address,
        value: "0",
        method: 0,
        gasPrice: "1",
        gasLimit: 1000,
        nonce: 0,
        params: messageHex
    }
}

export async function createLink (did: string, account: AccountID, provider: any, opts: BlockchainHandlerOpts): Promise<LinkProof> {
    const consentMessage = getConsentMessage(did, !opts.skipTimestamp)
    const addresses = await provider.getAccounts()
    const payload = asTransaction(addresses[0], JSON.stringify(consentMessage))
    const signatureResponse = await provider.sign(account.address, payload)
    const proof: LinkProof = {
        version: 2,
        type: 'eoa-tx',
        message: consentMessage.message,
        signature: signatureResponse.signature.data,
        account: account.toString()
    }
    if (!opts.skipTimestamp) proof.timestamp = consentMessage.timestamp
    return proof
}

export async function authenticate(message: string, account: AccountID, provider: any): Promise<string> {
    const addresses = await provider.getAccounts()
    const payload = asTransaction(addresses[0], JSON.stringify(message))
    const signatureResponse = await provider.sign(account.address, payload)
    return signatureResponse.signature.data
}

export async function validateLink (proof: LinkProof): Promise<LinkProof | null> {
    const signingTools = await import(moduleToImport)
    const account = new AccountID(proof.account)
    const consentMessage: ConsentMessage = {
        message: proof.message,
    }
    if (proof.timestamp) {
        consentMessage.timestamp = proof.timestamp
    }
    const payload = asTransaction(account.address, JSON.stringify(consentMessage))
    const transaction = signingTools.transactionSerialize(payload);
    const recover = signingTools.verifySignature(proof.signature, transaction);
    if (recover) {
        return proof
    } else {
        return null
    }
}

const Handler: BlockchainHandler = {
    namespace,
    authenticate,
    validateLink,
    createLink
}

export default Handler
