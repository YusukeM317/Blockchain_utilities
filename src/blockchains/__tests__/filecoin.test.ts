import {LocalFilecoinProvider} from "@ukstv/local-filecoin-provider";
import {authenticate, createLink, validateLink} from "../filecoin";
import {AccountID} from "caip";

const did = 'did:3:bafysdfwefwe'
const testnetPrivateKey = '7b2254797065223a22736563703235366b31222c22507269766174654b6579223a2257587362654d5176487a366f5668344b637262633045642b31362b3150766a6a504f3753514931355031343d227d'
const mainnetPrivateKey = '7b2254797065223a22736563703235366b31222c22507269766174654b6579223a2257587362654d5176487a366f5668344b637262633045642b31362b3150766a6a554f3753514931355031343d227d'
const testnetProvider = new LocalFilecoinProvider(testnetPrivateKey, true)
const mainnetProvider = new LocalFilecoinProvider(mainnetPrivateKey, false)

describe('createLink', () => {
    test('generate proof on testnet', async () => {
        const addresses = await testnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:t`)
        const proof = await createLink(did, account, testnetProvider, {skipTimestamp: true})
        expect(proof).toMatchSnapshot()
    })

    test('generate proof on mainnet', async () => {
        const addresses = await mainnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:f`)
        const proof = await createLink(did, account, mainnetProvider, {skipTimestamp: true})
        expect(proof).toMatchSnapshot()
    })

    test('fail on mainnet address for testnet provider', async () => {
        const addresses = await mainnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:f`)
        await expect(createLink(did, account, testnetProvider, {skipTimestamp: true})).rejects.toThrow()
    })
    test('fail on testnet address for mainnet provider', async () => {
        const addresses = await testnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:t`)
        await expect(createLink(did, account, mainnetProvider, {skipTimestamp: true})).rejects.toThrow()
    })
})

describe('validateLink', () => {
    test('testnet', async () => {
        const addresses = await testnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:t`)
        const proof = await createLink(did, account, testnetProvider, {skipTimestamp: true})
        await expect(validateLink(proof)).resolves.toEqual(proof)
    })

    test('mainnet', async () => {
        const addresses = await mainnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:f`)
        const proof = await createLink(did, account, mainnetProvider, {skipTimestamp: true})
        await expect(validateLink(proof)).resolves.toEqual(proof)

        const testAddr = await testnetProvider.getAccounts()
        const testAcc = new AccountID(`${testAddr[0]}@fil:t`)
        proof.account = testAcc.toString()
        await expect(validateLink(proof)).resolves.toEqual(null)
    })
})

describe('authenticate', () => {
    test('testnet', async () => {
        const addresses = await testnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:t`)
        expect(await authenticate('msg', account, testnetProvider)).toMatchSnapshot()
    })

    test('mainnet', async () => {
        const addresses = await mainnetProvider.getAccounts()
        const account = new AccountID(`${addresses[0]}@fil:m`)
        expect(await authenticate('msg', account, mainnetProvider)).toMatchSnapshot()
    })
})
