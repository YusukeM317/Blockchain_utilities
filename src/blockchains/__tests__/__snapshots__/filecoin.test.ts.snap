// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate mainnet 1`] = `"4yGIDAltS6TIG+8LJgrhxAJ7E8wkqWoKWfSH9JSEUpwlXiK1BMGIT1t1tP5kgkHZoMS+brFbkihtBCIVcChsTwE="`;

exports[`authenticate testnet 1`] = `"mZT9lhJBvrg+/elKOI/uwJWnsl/buGKZzNj2liujgqotMFVqOJ2KJVPFHvbvIJDbtpFjlcbBsAXtTeNwHePRnQE="`;

exports[`createLink generate proof on mainnet 1`] = `
Object {
  "account": "f1ymbkm43ysdiruibji6c6m2iy5bxwbuv5d6cdqmy@fil:f",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe",
  "signature": "ptn0BGzsCTD3tPjOtuCMhigE3A2BFadJkOahJ+AuPxNCCGSKTIfBVLb55QKPf0pl14+6YUnRAvdwk8lFc98aZAE=",
  "type": "eoa-tx",
  "version": 2,
}
`;

exports[`createLink generate proof on testnet 1`] = `
Object {
  "account": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka@fil:t",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe",
  "signature": "GuG1fGJQwLs1t33bp/esXhWCU3wIzENq9FBjntcQv2wNmk4Iu9D5KsnU3LpH8rGKPliQWHmkod0dOaJYrAS13QE=",
  "type": "eoa-tx",
  "version": 2,
}
`;
